File: src/srp/SrpHash.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace SecureRemotePassword
{
	/// <summary>
	/// Hashing algorithm for the SRP-6a protocol.
	/// </summary>
	public class SrpHash : ISrpHash
	{
		/// <summary>
		/// Initializes a new instance of the <see cref="SrpHash"/> class.
		/// </summary>
		/// <param name="hasher">The hashing algorithm.</param>
		/// <param name="algorithmName">The name of the algorithm.</param>
		public SrpHash(HashAlgorithm hasher, string algorithmName = null)
		{
			Hasher = hasher;
			AlgorithmName = algorithmName ?? Hasher.GetType().Name;
		}

		private HashAlgorithm Hasher { get; }

		/// <summary>
		/// Computes the hash of the specified <see cref="string"/> or <see cref="SrpInteger"/> values.
		/// </summary>
		/// <param name="values">The values.</param>
		public SrpInteger ComputeHash(params object[] values) =>
			ComputeHash(Combine(values.Select(v => GetBytes(v))));

		/// <summary>
		/// Gets the size of the hash in bytes.
		/// </summary>
		public int HashSizeBytes => Hasher.HashSize / 8;

		/// <summary>
		/// Gets the name of the algorithm.
		/// </summary>
		public string AlgorithmName { get; }

		private SrpInteger ComputeHash(byte[] data)
		{
			var hash = Hasher.ComputeHash(data);
			return SrpInteger.FromByteArray(hash);

			// should yield the same result:
			// var hex = hash.Aggregate(new StringBuilder(), (sb, b) => sb.Append(b.ToString("X2")), sb => sb.ToString());
			// return SrpInteger.FromHex(hex);
		}

		/// <summary>
		/// Creates the hasher for the given hashing algorithm.
		/// </summary>
		/// <param name="algorithm">The name of the hashing algorithm.</param>
		public static HashAlgorithm CreateHasher(string algorithm)
		{
			var result = default(HashAlgorithm);

			// CryptoConfig is not available on .NET Standard 1.6
			#if USE_CRYPTO_CONFIG
			result = (HashAlgorithm)CryptoConfig.CreateFromName(algorithm);
			#endif

			if (result == null)
			{
				switch (algorithm.ToLowerInvariant())
				{
					case "sha1":
						return SHA1.Create();

					case "sha256":
						return SHA256.Create();

					case "sha384":
						return SHA256.Create();

					case "sha512":
						return SHA256.Create();
				}
			}

			return result;
		}

		private static byte[] EmptyBuffer { get; } = new byte[0];

		private static byte[] GetBytes(object obj)
		{
			if (obj == null)
			{
				return EmptyBuffer;
			}

			var value = obj as string;
			if (!string.IsNullOrEmpty(value))
			{
				return Encoding.UTF8.GetBytes(value);
			}

			var integer = obj as SrpInteger;
			if (integer != null)
			{
				return integer.ToByteArray();
			}

			return EmptyBuffer;
		}

		private static byte[] Combine(IEnumerable<byte[]> arrays)
		{
			var rv = new byte[arrays.Sum(a => a.Length)];
			var offset = 0;

			foreach (var array in arrays)
			{
				Buffer.BlockCopy(array, 0, rv, offset, array.Length);
				offset += array.Length;
			}

			return rv;
		}
	}
}

File: src/srp-dart/srp-6a/lib/src/srp_hash.dart


import 'dart:convert';
import 'dart:typed_data';

import 'package:srp_6a/src/srp_types/srp_integer.dart';
import 'package:crypto/crypto.dart';

abstract class ISrpHash {
  /// Computes the hash of the specified [String] or [SrpInteger] values.
  SrpInteger computeHash(List<Object?> values);

  /// Gets the hash size in bytes.
  int get hashSizeBytes;

  /// Gets the name of the algorithm.
  String get algorithmName;
}

class SrpHash implements ISrpHash {
  final Hash Function() hasherFactory;
  final String? _algorithmName;

  SrpHash(this.hasherFactory, [this._algorithmName]);

  Hash get hasher => hasherFactory();

  @override
  SrpInteger computeHash(List<Object?> values) {
    final combined = _combine(values.map((v) => getBytes(v)).toList());
    return computeHashFromBytes(combined);
  }

  @override
  int get hashSizeBytes => hasher.convert(Uint8List(0)).bytes.length;

  @override
  String get algorithmName => _algorithmName ?? hasher.runtimeType.toString();

  SrpInteger computeHashFromBytes(Uint8List data) {
    final hash = hasher.convert(data).bytes;
    return SrpInteger.fromByteArray(Uint8List.fromList(hash));
  }

  static Hash createHasher(String algorithm) {
    switch (algorithm.toLowerCase()) {
      case 'md5':
        return md5;
      case 'sha1':
        return sha1;
      case 'sha256':
        return sha256;
      case 'sha384':
        return sha384;
      case 'sha512':
        return sha512;
      default:
        throw ArgumentError('Unknown algorithm: $algorithm');
    }
  }

  static Hash createSha256() {
    return sha256;
  }

  static Uint8List get emptyBuffer => Uint8List(0);

  static Uint8List getBytes(Object? obj) {
    if (obj == null) {
      return emptyBuffer;
    }

    if (obj is String && obj.isNotEmpty) {
      return Uint8List.fromList(utf8.encode(obj));
    }

    if (obj is SrpInteger) {
      return obj.toByteArray();
    }

    return emptyBuffer;
  }

  static Uint8List _combine(List<Uint8List> arrays) {
    final totalLength = arrays.fold(0, (sum, a) => sum + a.length);
    final result = Uint8List(totalLength);
    var offset = 0;

    for (final array in arrays) {
      result.setRange(offset, offset + array.length, array);
      offset += array.length;
    }

    return result;
  }
}

class SrpHashGeneric<T extends Hash> extends SrpHash {
  SrpHashGeneric() : super(() => createHasher(T.toString()));

  static Hash createHasher(String algorithm) {
    switch (algorithm.toLowerCase()) {
      case 'md5':
        return md5;
      case 'sha1':
        return sha1;
      case 'sha256':
        return sha256;
      case 'sha384':
        return sha384;
      case 'sha512':
        return sha512;
      default:
        throw ArgumentError('Unknown algorithm: $algorithm');
    }
  }
}